mongorestore/mongorestore.go

```diff
@@ -662,6 +673,74 @@ func (restore *MongoRestore) Restore() Result {
 	return result
 }
 
+// ReadPreludeMetadata finds and parses the prelude.json file if it's present.
+// It currently only sets the server.dumpServerVersion, but in the future we can read and set other metadata from the dump as required.
+// Returns true if the metadata file exists.
+func (restore *MongoRestore) ReadPreludeMetadata(target archive.DirLike) (bool, error) {
+	filename := "prelude.json"
+	if restore.InputOptions.Gzip {
+		filename += ".gz"
+	}
+
+	var reader io.ReadCloser
+	if !target.IsDir() {
+		target = target.Parent()
+	}
+	file, err := os.Open(filepath.Join(target.Path(), filename))
+	if errors.Is(err, os.ErrNotExist) {
+		// if prelude.json doesn't exist, check the parent directory in case db directory was used as target
+		file, err = os.Open(filepath.Join(target.Parent().Path(), filename))
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		} else if err != nil {
+			return false, err
+		}
+	}
+
+	defer file.Close()
+
+	if restore.InputOptions.Gzip {
+		zipfile, err := gzip.NewReader(file)
+		if err != nil {
+			return true, fmt.Errorf("failed to open gzip file %s: %w", filename, err)
+		}
+		defer zipfile.Close()
+		reader = zipfile
+	} else {
+		reader = file
+	}
+	bytes, err := io.ReadAll(reader)
+	if err != nil {
+		return true, fmt.Errorf("failed to read prelude metadata from %s: %w", filename, err)
+	}
+
+	var prelude map[string]string
+	err = json.Unmarshal(bytes, &prelude)
+	if err != nil {
+		return true, fmt.Errorf("failed to unmarshal prelude metadata from %s: %w", filename, err)
+	}
+
+	dumpVersion, ok := prelude["ServerVersion"]
+	if !ok {
+		return true, fmt.Errorf("ServerVersion key not found in %s", filename)
+	}
+
+	// mongodump sets server version to unknown if it can't get the server version
+	if dumpVersion == "unknown" {
```

Comment:
This should probably be a constant rather than a literal string that gets copied around.